File: 1049-last-stone-weight-ii/1049-last-stone-weight-ii.py
class Solution:
    def lastStoneWeightII(self, stones: List[int]) -> int:
        memo = [set() for _ in range(len(stones))]
        
        memo[0].add(stones[0])
        
        for i in range(1, len(stones)):
            curr = stones[i]
            
            for prev in memo[i - 1]:
                p1 = abs(prev + curr)
                p2 = abs(prev - curr)
                
                memo[i].add(p1)
                memo[i].add(p2)
                
        return min(memo[-1])

File: 1049-last-stone-weight-ii/1049-last-stone-weight-ii.cs
public class Solution {
    public int LastStoneWeightII(int[] stones) {
        
        HashSet<int>[] dp = new HashSet<int>[stones.Length];
        dp[0] = new HashSet<int>();
        dp[0].Add(stones[0]);
      
        for(int i = 1; i < stones.Length; i++){
            dp[i] = new HashSet<int>();
            int curr = stones[i];
            
            foreach(int sum in dp[i - 1]){
                dp[i].Add(Math.Abs(curr - sum));
                dp[i].Add(Math.Abs(curr + sum));
            };
        }
        
        return dp[stones.Length - 1].Min();
       
    }
}